### PR TITLE
Update to keep track of superseded products to avoid false warnings

### LINKED ISSUE #597: Validate does not maintain history of other versions it comes across as it traverses directories causing erroneous WARNING messages
### Checked for duplicates

Yes - I have checked

### 🐛 Describe the bug

Per https://github.com/NASA-PDS/operations/issues/322#issuecomment-1377978726 and the validate report denoted in https://github.com/NASA-PDS/operations/issues/322, validate is throwing erroneous WARNING messages for past versions of bundles/collections in the directory used as the target.

### 🕵️ Expected behavior

Validate would maintain an understanding of past versions of bundles and the past versions of collections they refer to, and simple SKIP validation of those products without throwing a WARNING.

### 📜 To Reproduce

1. download either of the test data sets
2. run `validate -R pds4.bundle -t /path/to/naif/dataset`
```
...
  PASS: file:/Users/jpadams/test/naif3/spice_kernels/collection_spice_kernels_inventory_v001.csv
      WARNING  [warning.file.not_referenced_in_label]   File is not referenced by any label
        3 integrity check(s) completed

  PASS: file:/Users/jpadams/test/naif3/spice_kernels/collection_spice_kernels_inventory_v002.csv
      WARNING  [warning.file.not_referenced_in_label]   File is not referenced by any label
        4 integrity check(s) completed
...
Summary:

  0 error(s)
  2 warning(s)

  Product Validation Summary:
    17         product(s) passed
    0          product(s) failed
    4          product(s) skipped

  Referential Integrity Check Summary:
    19         check(s) passed
    0          check(s) failed
    0          check(s) skipped

  Message Types:
    2            warning.file.not_referenced_in_label
```

### 🖥 Environment Info

- Version of this software: 3.1.1
- Operating System: MacOSX

### 📚 Version of Software Used

v3.1.1

### 🩺 Test Data / Additional context

https://naif.jpl.nasa.gov/pub/naif/pds/pds4/mars2020/mars2020_spice/

or for a smaller data set:

https://pds.nasa.gov/data/pds4/test-data/custom-datasets/naif3/

### 🦄 Related requirements

_No response_

### ⚙️ Engineering Details

Looks like regression was introduced prior to v3.1.1. Do not see this in v3.0.3

### PR DESCRIPTION
**Pull request by al-niessner**
## 🗒️ Summary
Not the most comprehensive XML parsing but the shortest road to fix the bug. It can be cleverly circumvented with XML comments. To fix this, if it ever becomes and issue, is to introduce XML DOM parsing and XPATH for finding the tags. Unnecessary work for cooperative opponents but required for adversarial opponents.

## ⚙️ Test Data and/or Report

See github597 in unit testing.

## ♻️ Related Issues
Closes #597

### CODE REVIEW

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FindUnreferencedFiles.java [hunk 1]
```diff
@@ -26,13 +32,46 @@
...
+          begin = content.indexOf(openTag, end + closeTag.length());
+        }
+      } catch (IOException e) {
+        // TODO Auto-generated catch block
+        e.printStackTrace();
+      } finally {
+        if (uis != null) {
+          try {
+            uis.close();
+          } catch (IOException e) {}
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FindUnreferencedFiles.java:60]: @sonatype-lift ignore

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FindUnreferencedFiles.java [hunk 2]
```diff
@@ -26,13 +32,46 @@
...
+        begin = content.indexOf(openTag);
+        
+        while (-1 < begin) {
+          end = content.indexOf(closeTag, begin);
+          referencedBySkipped.add("file:" + Paths.get(path, content.substring(begin+openTag.length(), end)).toString());
+          begin = content.indexOf(openTag, end + closeTag.length());
+        }
+      } catch (IOException e) {
+        // TODO Auto-generated catch block
+        e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FindUnreferencedFiles.java:55]: @sonatype-lift ignore

pre existing

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FindUnreferencedFiles.java [hunk 3]
```diff
@@ -26,13 +32,46 @@
...
-
+  private HashSet<String> buildReferencedFromSkipped() {
+    HashSet<String> referencedBySkipped = new HashSet<String>();
+    InputStream uis = null;
+    final String closeTag = "</file_name>", openTag = "<file_name>";
+    for (URL skipped : SkippedItems.getInstance().copy()) {
+      try {
+        int begin=0,end=-1;
+        uis = skipped.openStream();
+        String content = new String(uis.readAllBytes());
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FindUnreferencedFiles.java]: fixed

File: src/main/java/gov/nasa/pds/tools/validate/SkippedItems.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package gov.nasa.pds.tools.validate;
+
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+
+public class SkippedItems {
+  final private HashSet<URL> skipped = new HashSet<URL>();
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply src/main/java/gov/nasa/pds/tools/validate/SkippedItems.java]: fixed

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FindUnreferencedFiles.java [hunk 5]
```diff
@@ -26,13 +32,46 @@
...
+            uis.close();
+          } catch (IOException e) {}
+          uis = null;
+        }
+      }
+    }
+    return referencedBySkipped;
+  }
+  private String extractPath (URL url) {
+      Path fullpath = Paths.get(url.getPath());
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FindUnreferencedFiles.java:68]: @sonatype-lift ignore

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FindUnreferencedFiles.java [hunk 6]
```diff
@@ -26,13 +32,46 @@
...
+      try {
+        int begin=0,end=-1;
+        uis = skipped.openStream();
+        String content = new String(uis.readAllBytes());
+        String path = this.extractPath(skipped);
+        begin = content.indexOf(openTag);
+        
+        while (-1 < begin) {
+          end = content.indexOf(closeTag, begin);
+          referencedBySkipped.add("file:" + Paths.get(path, content.substring(begin+openTag.length(), end)).toString());
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FindUnreferencedFiles.java:50]: @sonatype-lift ignore

File: src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FindUnreferencedFiles.java [hunk 7]
```diff
@@ -26,13 +32,46 @@
...
   private static final Logger LOG = LoggerFactory.getLogger(FindUnreferencedIdentifiers.class);
-
+  private HashSet<String> buildReferencedFromSkipped() {
+    HashSet<String> referencedBySkipped = new HashSet<String>();
+    InputStream uis = null;
+    final String closeTag = "</file_name>", openTag = "<file_name>";
+    for (URL skipped : SkippedItems.getInstance().copy()) {
+      try {
+        int begin=0,end=-1;
+        uis = skipped.openStream();
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FindUnreferencedFiles.java:43]: @sonatype-lift ignore

File: src/main/java/gov/nasa/pds/tools/validate/SkippedItems.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+public class SkippedItems {
+  final private HashMap<String,URL> skipped = new HashMap<String,URL>();
+  private static SkippedItems singleton = null;
+  private SkippedItems() {
+  }
+  public static synchronized SkippedItems getInstance() {
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/tools/validate/SkippedItems.java:13]: I like this. We need to see more "design patterns"!

**Review by nutjob4life — APPROVED** [review]
👍

**Pull request merged**